build.py

```diff
@@ -0,0 +1,143 @@
+
+
+name = "pysparkling"
+default_task = "publish"
+version = _get_pysparkling_version()
+license = 'MIT'
+summary = 'Pure Python implementation of the Spark RDD interface.'
+description = None
+authors = [
+    Author(name='Sven Kreiss', email='me@svenkreiss.com'),
+    Author(name="Erwan Guyomarc'h", email='tools4origins@gmail.com'),
+    Author(name='Steven Van Ingelgem', email='steven@vaningelgem.be'),
+]
+maintainers = [
+    Author(name="Erwan Guyomarc'h", email='tools4origins@gmail.com'),
+]
+requires_python = '>= 3.4'
+url = 'https://github.com/tools4origins/pysparkling'
+
+_logging_root_level: int = logging.root.getEffectiveLevel()
+
+
+@before('run_unit_tests')
+def _set_debug_mode():
+    logging.root.setLevel('DEBUG')
+
+
+@after('run_unit_tests')
+def _set_debug_mode_after():
+    logging.root.setLevel(_logging_root_level)
+
+
+@after('package')
+def _add_extras_require(project, logger):
+    indent_size = 4
+    encoding = 'utf-8'
+
+    setup_script = Path(project.expand_path("$dir_dist", "setup.py"))
+    logger.info("Adding 'extras_require' to setup.py")
+    setup = setup_script.read_text(encoding=encoding).rstrip()
+    if setup[-1] != ')':
+        raise BuildFailedException("This setup.py seems to be wrong?")
+
+    # Get the requirements-dev.txt file line by line, ready for insertion.
+    requirements_dev = '\n'.join(
+        ' '*4*indent_size + "'" + x.strip() + "',"
+        for x in (Path(__file__).parent / 'requirements-build.txt').read_text().split('\n')
+        if x
+    )
+
+    # TODO: find a nicer way to embed this!
+    new_setup = (
+        setup[:-1].rstrip()
+        + f"""
+        extras_require={{
+            'hdfs': ['hdfs>=2.0.0'],
+            'pandas': ['pandas>=0.23.2'],
+            'performance': ['matplotlib>=1.5.3'],
+            'streaming': ['tornado>=4.3'],
+            'test': [
+{requirements_dev}
+            ]
+        }},
+    )
+"""
+    )
+
+    setup_script.write_text(new_setup, encoding=encoding)
+
+
+@init
+def set_properties(project):
+    # Small tweak to project.list_scripts() as that method lists EVERYTHING in the scripts directory.
+    #   and we're only interested in *.py files:
+    old_project_list_scripts = project.list_scripts
+
+    def _my_list_scripts():
+        return [
+            filename
+            for filename in old_project_list_scripts()
+            if filename.lower().endswith('.py')
+        ]
+    setattr(project, 'list_scripts', _my_list_scripts)
+
+    project.set_property('dir_source_main_python', 'src/')
+    project.set_property('dir_source_main_scripts', 'scripts/')
+
+    project.depends_on_requirements(file='requirements.txt')
+
+    project.set_property('distutils_readme_description', True)
+    project.set_property('distutils_readme_file', 'README.rst')
+
+    project.set_property_if_unset("pytest_extra_args", [])
+    project.get_property("pytest_extra_args").append("-x")  # Fail on first failing unittest
```

Comment:
Is this new? I think it is useful to know if an error is found in only one test or affecting a lot of them when you start debugging

Reply:
Subject for a discussion. I like the fail fast option. But it depends on the situation.
For example at work we have tests that take 2h to go through (don't ask!), so it's better to fail fast.

I'll remove this.